pkg/tfbridge/provider.go

```diff
@@ -949,12 +952,35 @@ func (p *Provider) Diff(ctx context.Context, req *pulumirpc.DiffRequest) (*pulum
 
 	ic := newIgnoreChanges(ctx, schema, fields, olds, news, req.GetIgnoreChanges())
 
+	// if we have type errors that were generated during check
+	// we don't want to log the panic. In the future these type errors
+	// will hard fail during check and we will never make it to create
+	//
+	// We do our best to restrict catching panics to where the panic will
+	// occur. The type checking will catch panics that occur as part
+	// of the `Diff` function
+	recoverFunc := func(r interface{}) {
+		if _, ok := p.hasTypeErrors[urn]; !ok {
+			panic(r)
+		}
+	}
+	defer func() {
+		if r := recover(); r != nil {
+			recoverFunc(r)
+		}
+	}()
```

Comment:
We can refactor this into a closure instead of a mutable cleanup operation.
```go
func callWithRecover[T any](rec func(any) error, call func() (T, error)) (_ T, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = rec(r)
		}
	}()
	
	return call()
}

// if we have type errors that were generated during check
// we don't want to log the panic. In the future these type errors
// will hard fail during check and we will never make it to create
//
// We do our best to restrict catching panics to where the panic will
// occur. The type checking will catch panics that occur as part
// of the `Diff` function
func (p *Provider) recoverOnTypeError(payload any) error {
        // TODO: Maybe we should recover the stack trace here?
	if _, ok := p.hasTypeErrors[urn]; !ok {
			panic(r)
	}
	if err, ok := payload.(error); ok {
		return fmt.Errorf("panicked: %w", err)
        }
	return fmt.Errorf("panicked: %#v", payload)
}
```

Usage would then look like this:

```go
diff, err := callWithRecover(p.recoverOnTypeError, func() (shim.DiffResult, error) {
	return p.tf.Diff(ctx, res.TFName, nil, config, shim.DiffOptions{
		TimeoutOptions: shim.TimeoutOptions{
			ResourceTimeout:  timeouts,
			TimeoutOverrides: newTimeoutOverrides(shim.TimeoutCreate, req.Timeout),
		},
	})
}
```